Intermittent failures in ReactiveStreamsPublisherTckTest.stochastic_spec103_mustSignalOnMethodsSequentially

As discussed in https://issues.redhat.com/browse/WFLY-19791, when running the MicroProfile REST Client TCK with WildFly we are seeing intermittent failures in ReactiveStreamsPublisherTckTest.stochastic_spec103_mustSignalOnMethodsSequentially. Failures result in this logging:

```
------- Stderr: -------
[AbstractSseTest] [ERROR] launchServer caughtException 
java.lang.RuntimeException: Failed to start SSE HTTP server
	at org.eclipse.microprofile.rest.client.tck.sse.HttpSseServer.start(HttpSseServer.java:46)
	at org.eclipse.microprofile.rest.client.tck.sse.AbstractSseTest.lambda$launchServer$0(AbstractSseTest.java:48)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.io.IOException: Failed to bind to ::/[0:0:0:0:0:0:0:0]:10000
	at org.eclipse.jetty.server.ServerConnector.openAcceptChannel(ServerConnector.java:349)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:310)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:234)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
	at org.eclipse.jetty.server.Server.doStart(Server.java:401)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
	at org.eclipse.microprofile.rest.client.tck.sse.HttpSseServer.start(HttpSseServer.java:43)
	... 6 more
Caused by: java.net.BindException: Address already in use
	at java.base/sun.nio.ch.Net.bind0(Native Method)
	at java.base/sun.nio.ch.Net.bind(Net.java:555)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.netBind(ServerSocketChannelImpl.java:337)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:294)
	at java.base/sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:89)
	at org.eclipse.jetty.server.ServerConnector.openAcceptChannel(ServerConnector.java:344)
	... 13 more
[ReactiveStreamsPublisherTckTest] [ERROR] Failed to create publisher
java.lang.AssertionError: Mock Sse Server did not start as expected expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.failNotEquals(Assert.java:1413)
	at org.testng.Assert.assertTrue(Assert.java:56)
	at org.eclipse.microprofile.rest.client.tck.sse.AbstractSseTest.launchServer(AbstractSseTest.java:62)
	at org.eclipse.microprofile.rest.client.tck.sse.ReactiveStreamsPublisherTckTest.createPublisher(ReactiveStreamsPublisherTckTest.java:106)
	at org.reactivestreams.tck.PublisherVerification.activePublisherTest(PublisherVerification.java:1134)
	at org.reactivestreams.tck.PublisherVerification$5.apply(PublisherVerification.java:251)
	at org.reactivestreams.tck.PublisherVerification$5.apply(PublisherVerification.java:248)
	at org.reactivestreams.tck.PublisherVerification.stochasticTest(PublisherVerification.java:1204)
	at org.reactivestreams.tck.PublisherVerification.stochastic_spec103_mustSignalOnMethodsSequentially(PublisherVerification.java:248)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:658)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:219)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:923)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:192)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:808)
	at org.testng.TestRunner.run(TestRunner.java:603)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:429)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:423)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:383)
	at org.testng.SuiteRunner.run(SuiteRunner.java:326)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1249)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
	at org.testng.TestNG.runSuites(TestNG.java:1092)
	at org.testng.TestNG.run(TestNG.java:1060)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeMulti(TestNGDirectoryTestSuite.java:193)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:94)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
java.lang.AssertionError: Mock Sse Server did not start as expected expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.failNotEquals(Assert.java:1413)
	at org.testng.Assert.assertTrue(Assert.java:56)
	at org.eclipse.microprofile.rest.client.tck.sse.AbstractSseTest.launchServer(AbstractSseTest.java:62)
	at org.eclipse.microprofile.rest.client.tck.sse.ReactiveStreamsPublisherTckTest.createPublisher(ReactiveStreamsPublisherTckTest.java:106)
	at org.reactivestreams.tck.PublisherVerification.activePublisherTest(PublisherVerification.java:1134)
	at org.reactivestreams.tck.PublisherVerification$5.apply(PublisherVerification.java:251)
	at org.reactivestreams.tck.PublisherVerification$5.apply(PublisherVerification.java:248)
	at org.reactivestreams.tck.PublisherVerification.stochasticTest(PublisherVerification.java:1204)
	at org.reactivestreams.tck.PublisherVerification.stochastic_spec103_mustSignalOnMethodsSequentially(PublisherVerification.java:248)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:658)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:219)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:923)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:192)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:808)
	at org.testng.TestRunner.run(TestRunner.java:603)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:429)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:423)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:383)
	at org.testng.SuiteRunner.run(SuiteRunner.java:326)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1249)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
	at org.testng.TestNG.runSuites(TestNG.java:1092)
	at org.testng.TestNG.run(TestNG.java:1060)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeMulti(TestNGDirectoryTestSuite.java:193)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:94)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

I believe the issue is stochastic_spec103_mustSignalOnMethodsSequentially is a stochastic test, which means its test code runs multiple (i.e. 100) times inside a single @Test method. This breaks the design of the @AfterMethod ReactiveStreamsPulisherTckTest.countDownLatch() method, which is meant to block until the 'cleanupLatch' created in createPublisher() has tripped. The effect is as the stochastic test loops, there's no guarantee that the HttpSseServer spawned in createPublisher by one iteration has completed shutdown before the next iteration begins.

I'm not sure what the fix would be since ReactiveStreamsPublisherTckTest depends on the external PublisherVerificationTest for so much.

Perhaps  ReactiveStreamsPublisherTckTest.createPublisher, before creating a new CountDownLatch and assigning to 'cleanupLatch', could check if cleanupLatch is not null and if not await it.

--- a/tck/src/main/java/org/eclipse/microprofile/rest/client/tck/sse/ReactiveStreamsPublisherTckTest.java
+++ b/tck/src/main/java/org/eclipse/microprofile/rest/client/tck/sse/ReactiveStreamsPublisherTckTest.java
@@ -102,6 +102,16 @@ public class ReactiveStreamsPublisherTckTest extends PublisherVerification<Inbou
 
         CountDownLatch stopLatch = new CountDownLatch(1);
         try {
+            if (cleanupLatch != null) {
+                // We null out cleanupLatch in a @BeforeTest method, so if it is not null
+                // we must be in a test that calls this multiple times in a single @Test method,
+                // e.g. a stochastic test. In this case we can't rely on having awaited cleanupLatch
+                // in our @AfterTest method, and we need to do it ourselves.
+                if (!cleanupLatch.await(30, TimeUnit.SECONDS)) {
+                    // Just log; maybe by the time we create a new server the old one will be stopped.
+                    LOG.error("Server did not close long after test completed");
+                }
+            }
             cleanupLatch = new CountDownLatch(1);
             AtomicReference<Throwable> serverException = launchServer(stopLatch, es -> {
                 for (long i = 0; i < elements; i++) {

